File: packages/shared-ts/src/form-body-unmask.ts
import type { SecretsPipeline } from './secrets-pipeline.js';

export type FormBodyUnmasker = Pick<SecretsPipeline, 'unmaskBody' | 'hasSecrets'>;

function decodeFormComponent(raw: string): string | null {
  try {
    return decodeURIComponent(raw.replace(/\+/g, '%20'));
  } catch {
    return null;
  }
}

function encodeFormComponent(value: string): string {
  return encodeURIComponent(value);
}

export function unmaskFormBody(
  pipeline: FormBodyUnmasker,
  body: string,
  hostname: string
): { text: string } {
  if (!body || !pipeline.hasSecrets()) return { text: body };

  let changed = false;
  const fields = body.split('&').map((field) => {
    const eq = field.indexOf('=');
    const name = eq < 0 ? '' : field.slice(0, eq + 1);
    const rawValue = eq < 0 ? field : field.slice(eq + 1);
    if (!rawValue) return field;

    const decoded = decodeFormComponent(rawValue);
    if (decoded === null) {
      const { text } = pipeline.unmaskBody(rawValue, hostname);
      if (text === rawValue) return field;
      changed = true;
      return `${name}${text}`;
    }

    const { text } = pipeline.unmaskBody(decoded, hostname);
    if (text === decoded) return field;
    changed = true;
    return `${name}${encodeFormComponent(text)}`;
  });

  return { text: changed ? fields.join('&') : body };
}

File: packages/swift-server/Sources/Server/FormBodyUnmask.swift
import Foundation





















private let formComponentAllowed = CharacterSet(
    charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_.!~*'()"
)




private func decodeFormComponent(_ raw: String) -> String? {
    raw.replacingOccurrences(of: "+", with: "%20").removingPercentEncoding
}



private func encodeFormComponent(_ value: String) -> String {
    value.addingPercentEncoding(withAllowedCharacters: formComponentAllowed) ?? value
}



func unmaskFormBody(text body: String, hostname: String, injector: SecretInjector) -> String {
    if body.isEmpty || injector.isEmpty { return body }

    var changed = false
    let fields = body.components(separatedBy: "&").map { field -> String in
        let name: String
        let rawValue: String
        if let eq = field.firstIndex(of: "=") {
            name = String(field[field.startIndex...eq])
            rawValue = String(field[field.index(after: eq)...])
        } else {
            name = ""
            rawValue = field
        }
        if rawValue.isEmpty { return field }

        guard let decoded = decodeFormComponent(rawValue) else {
            
            
            
            
            let replaced = injector.injectBody(text: rawValue, hostname: hostname)
            if replaced == rawValue { return field }
            changed = true
            return name + replaced
        }

        let replaced = injector.injectBody(text: decoded, hostname: hostname)
        if replaced == decoded { return field }
        changed = true
        return name + encodeFormComponent(replaced)
    }

    return changed ? fields.joined(separator: "&") : body
}
